metals/src/main/scala/scala/meta/internal/metals/StacktraceAnalyzer.scala

```diff
@@ -0,0 +1,227 @@
+package scala.meta.internal.metals
+
+import java.io.FileWriter
+import java.net.URLEncoder
+import java.util.Collections.singletonList
+
+import scala.io.Source
+import scala.util.Try
+
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.io.AbsolutePath
+
+import com.google.gson.JsonPrimitive
+import org.eclipse.lsp4j.Location
+import org.eclipse.{lsp4j => l}
+
+class StacktraceAnalyzer(
+    workspace: AbsolutePath,
+    definitionProvider: DefinitionProvider,
+    gotoicon: String,
+    commandsInHtmlSupported: Boolean
+) {
+
+  def analyzeCommand(
+      commandParams: l.ExecuteCommandParams
+  ): Option[l.ExecuteCommandParams] = {
+    parseJsonParams(commandParams)
+      .flatMap(analyzeStackTrace)
+  }
+
+  def matches(path: AbsolutePath): Boolean =
+    path == workspace.resolve(Directories.stacktrace)
+
+  def highlight(
+      path: AbsolutePath,
+      lineNumber: Int
+  ): java.util.List[l.DocumentHighlight] = {
+    val lineOpt = readStacktraceFileLine(path, lineNumber)
+    lineOpt match {
+      case Some(line) =>
+        val from = new l.Position(lineNumber, line.indexOf("at ") + 3)
+        val to = new l.Position(lineNumber, line.indexOf("("))
+        val range = new l.Range(from, to)
+        singletonList(
+          new l.DocumentHighlight(range, l.DocumentHighlightKind.Read)
+        )
+      case None =>
+        java.util.Collections.emptyList()
+    }
+  }
+
+  def stacktraceLenses(path: AbsolutePath): Seq[l.CodeLens] = {
+    stacktraceLenses(readStacktraceFile(path))
+  }
+
+  private def readStacktraceFileLine(
+      path: AbsolutePath,
+      lineNumber: Int
+  ): Option[String] =
+    readStacktraceFile(path).drop(lineNumber).headOption
+
+  private def readStacktraceFile(path: AbsolutePath): List[String] = {
+    Source.fromFile(path.toString).getLines().toList
```

Comment:
This should be in buffers, no need to read from Source.

Reply:
changed